packages/@sanity/base/src/actions/utils/types.ts

```diff
@@ -1,52 +1,94 @@
+import {SanityDocument} from '@sanity/types'
 import React from 'react'
 
 export interface ActionComponent<ActionProps> {
-  (props: ActionProps): ActionDescription
-}
-
-interface Document {
-  _id: string
-  _type: string
+  (props: ActionProps): ActionDescription | null
 }
 
 export interface DocumentActionProps {
   id: string
   type: string
-  draft: null | Document
-  published: null | Document
+  draft: SanityDocument | null
+  liveEdit: boolean
+  published: SanityDocument | null
+  revision: string
   onComplete: () => void
 }
 
 export type DocumentActionComponent = ActionComponent<DocumentActionProps>
 
-interface ConfirmDialogProps {
+export interface DocumentActionLegacyDialogProps {
+  type: 'legacy'
+  content: React.ReactNode
+}
+
+export interface DocumentActionConfirmDialogProps {
   type: 'confirm'
   color?: 'warning' | 'success' | 'danger' | 'info'
   message: React.ReactNode
   onConfirm: () => void
   onCancel: () => void
 }
 
-// Todo: move these to action spec/core types
-interface ModalDialogProps {
+export interface DocumentActionModalDialogProps {
   type: 'modal'
   content: React.ReactNode
+  /**
+   * @beta
+   */
+  footer?: React.ReactNode
+  /**
+   * @beta
+   */
+  header?: React.ReactNode
   onClose: () => void
+  showCloseButton?: boolean
+  /**
+   * @beta
+   */
+  width?: 'small' | 'medium' | 'large' | 'full'
 }
 
-// Todo: move these to action spec/core types
-interface PopOverDialogProps {
+export interface DocumentActionPopOverDialogProps {
   type: 'popover'
   content: React.ReactNode
   onClose: () => void
 }
 
+/**
+ * @deprecated
+ */
+export interface DocumentActionErrorDialogProps {
+  type: 'error'
+  content?: React.ReactNode
+  onClose: () => void
+  title?: React.ReactNode
+}
+
+/**
+ * @deprecated
+ */
+export interface DocumentActionSuccessDialogProps {
+  type: 'success'
+  content?: React.ReactNode
+  onClose: () => void
+  title?: React.ReactNode
+}
+
+export type DocumentActionDialogProps =
+  | DocumentActionLegacyDialogProps
+  | DocumentActionConfirmDialogProps
+  | DocumentActionPopOverDialogProps
+  | DocumentActionModalDialogProps
+  | DocumentActionErrorDialogProps
+  | DocumentActionSuccessDialogProps
+
 export interface ActionDescription {
   label: string
   icon?: React.ReactNode
   disabled?: boolean
-  shortcut?: string
-  title?: string
-  dialog?: ConfirmDialogProps | PopOverDialogProps | ModalDialogProps
+  shortcut?: string | null
+  title?: React.ReactNode
+  dialog?: DocumentActionDialogProps | false | null
```

Comment:
I added `false` here because our document actions will pass the type-checker then:

We sometimes do:
```ts
const dialog = dialogOpen && {type: 'modal', ...}
```

In this case, the dialog is either false or an object.

Reply:
i think you're right here. we have an example of this in the docs too